### ISSUE TITLE
S3 putBucketReplication fails with SignatureDoesNotMatch

### ISSUE THREAD

**Issue by jgilbert01** [aws/aws-sdk-js#910]
I have the following code and all the calls to s3 succeed except the last call to putBucketReplication.
I know my key is correct, because the other calls work.
I have searched high and low for clues on how this fix this, but I am out of ideas.

I welcome a fresh set of eyes on this.

sdk version = 2.2.38
node version = v0.12.4

`
        var s3 = Promise.promisifyAll(new aws.S3({
            httpOptions: { timeout: 1000 },
            logger: process.stdout,
            credentials: new aws.SharedIniFileCredentials({ profile: account.profile })
        }));

        return s3.createBucketAsync({
            Bucket: bucketConfig.name,
            CreateBucketConfiguration: bucketConfig.createBucketConfiguration
        })
            .then(function (data) {
                    return s3.putBucketPolicyAsync({
                        Bucket: bucketConfig.name,
                        Policy: JSON.stringify(bucketConfig.policy)
                    });
            })
            .then(function (data) {
                    return s3.putBucketVersioningAsync({
                        Bucket: bucketConfig.name,
                        VersioningConfiguration: {
                            Status: 'Enabled'
                        }
                    });
            })
            .then(function (data) {
                    return s3.putBucketLifecycleConfigurationAsync({
                        Bucket: bucketConfig.name,
                        LifecycleConfiguration: bucketConfig.lifecycleConfiguration
                    });
            })
            .then(function (data) {
                    return s3.putBucketReplicationAsync({
                        Bucket: bucketConfig.name,
                        ReplicationConfiguration: bucketConfig.replicationConfiguration
                    });
            })
`

The putBucketReplication params look like this:

`
{

    Bucket: 'my-bucket',
    ReplicationConfiguration: {
        Role: 'arn:aws:iam::0123456789:role/my-bucket-rep-role',
        Rules: [ {
                Prefix: '',
                Status: 'Enabled',
                Destination: {
                    Bucket: 'my-other-bucket',
                    StorageClass: 'REDUCED_REDUNDANCY'
                }
            }
        ]
    }
}
`

**Comment by chrisradek** [comment]
@jgilbert01 
Can you share what the error is you're getting when calling `putBucketReplication`?

**Comment by jgilbert01** [comment]
```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><AWSAccessKeyId>abc</AWSAccessKeyId><StringToSign>PUT
PMNCvA+qhZBX4gjojTebeQ==
application/octet-stream

x-amz-date:Wed, 24 Feb 2016 20:38:17 GMT
/dante-dev-backup-test/?replication</StringToSign><SignatureProvided>w2GjyBFNcg31oCsfMu0eDNApOos=</SignatureProvided><StringToSignBytes>50 55 54 0a 50 4d 4e 43 76 41 2b 71 68 5a 42 58 34 67 6a 6f 6a 54 65 62 65 51 3d 3d 0a 61 70 70 6c 69 63 61 74 69 6f 6e 2f 6f 63 74 65 74 2d 73 74 72 65 61 6d 0a 0a 78 2d 61 6d 7a 2d 64 61 74 65 3a 57 65 64 2c 20 32 34 20 46 65 62 20 32 30 31 36 20 32 30 3a 33 38 3a 31 37 20 47 4d 54 0a 2f 64 61 6e 74 65 2d 64 65 76 2d 62 61 63 6b 75 70 2d 74 65 73 74 2f 3f 72 65 70 6c 69 63 61 74 69 6f 6e</StringToSignBytes><RequestId>5B455EB59D581317</RequestId><HostId>ocutt36TT9zs3r5iC4P0a5WBLw2hZRERpVHtir5nEot9Jm3rO7+kJ78+ftm8Scdh7Ql+bcZzMFw=</HostId></Error>
```

**Comment by chrisradek** [comment]
@jgilbert01 
Thanks for the error message. Depending on which signature version I configure the S3 client to use, I am getting different error messages. I haven't pinpointed where the issue is when using the v2 signer yet.

It looks like `ContentMD5` is actually required for `putBucketReplication`. With the v4 signer, `ContentMD5` isn't calculated unless it is marked as required (which it isn't currently in the SDK.) This is something that we'll have to address in the SDK, but there is a temporary workaround you can do until a fix is in place.

```javascript
//Set the S3 client to use signature version 4
var s3 = new AWS.S3({
  signatureVersion: 'v4'
});

//Grab a reference to the AWS.Request object
var req = s3.putBucketReplication(params);

//Workaround - set the Content-MD5 header
req.on('build', function(req) {
  var md5 = AWS.util.crypto.md5(req.httpRequest.body, 'base64');
  req.httpRequest.headers['Content-MD5'] = md5;
});

//Trigger the service call
req.send(function(err, data) {
  if (err) {
    console.log(err);
  } else {
    console.log(data);
  }
});
```

One more 'gotcha'. This isn't pointed out in our SDK API docs, but the Destinaton bucket needs to be the bucket ARN, not the name. This wasn't obvious until I looked at the S3 docs:
http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTreplication.html

So, instead of specifying the bucket name, you'd specify the ARN using the following pattern:
```
arn:aws:s3:::bucket_name
```

**Comment by jgilbert01** [comment]
Excellent!
The workaround does the trick.
Thx!

**Comment by chrisradek** [comment]
@jgilbert01 
I merged in the fix for this with pr #924 
It'll be available in the next version of the SDK, or can be installed from this git repo.
Thanks for reporting it!
